File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        array = []
        num = 1
        # left to right
        for i in nums:
            array.append([num])
            num *= i
        
        num = 1
        # right to left
        for idx, i in enumerate(nums[::-1]):
            array[len(nums) - 1 - idx].append(num)
            num *= i
        
        result = []

        for value in array:
            result.append(value[0] * value[1])
        
        return result

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.cpp
class Solution {
   public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;

        // If the tree is empty, return an empty result.
        if (!root) return res;

        // Queue for performing level-order (BFS) traversal.
        queue<TreeNode*> q;
        q.push(root);

        // Traverse the tree level by level.
        while (!q.empty()) {
            // Number of nodes present at the current level.
            int levelSize = q.size();

            // Process all nodes of the current level.
            for (int i = 0; i < levelSize; i++) {
                // Remove the front node from the queue.
                TreeNode* node = q.front();
                q.pop();

                // Since we process nodes from left to right,
                // the last node at this level is the rightmost node.
                if (i == levelSize - 1) res.push_back(node->val);

                // Add the left child to the queue.
                if (node->left) q.push(node->left);

                // Add the right child to the queue.
                if (node->right) q.push(node->right);
            }
        }

        // Return the list of nodes visible from the right side.
        return res;
    }
};